src/Evolution/DiscontinuousGalerkin/Messages/BoundaryMessage.cpp

```diff
@@ -0,0 +1,216 @@
+BoundaryMessage<Dim>::BoundaryMessage(
+    const size_t subcell_ghost_data_size_in, const size_t dg_flux_data_size_in,
+    const bool sent_across_nodes_in,
+    const ::TimeStepId& current_time_step_id_in,
+    const ::TimeStepId& next_time_step_id_in,
+    const Mesh<Dim>& volume_or_ghost_mesh_in,
+    const Mesh<Dim - 1>& interface_mesh_in, double* subcell_ghost_data_in,
+    double* dg_flux_data_in)
+    : subcell_ghost_data_size(subcell_ghost_data_size_in),
+      dg_flux_data_size(dg_flux_data_size_in),
+      sent_across_nodes(sent_across_nodes_in),
+      current_time_step_id(current_time_step_id_in),
+      next_time_step_id(next_time_step_id_in),
+      volume_or_ghost_mesh(volume_or_ghost_mesh_in),
+      interface_mesh(interface_mesh_in),
+      subcell_ghost_data(subcell_ghost_data_in),
+      dg_flux_data(dg_flux_data_in) {}
+
+template <size_t Dim>
+size_t BoundaryMessage<Dim>::total_size_without_data() {
+  // subcell_ghost_data_size, dg_flux_data_size
+  size_t totalsize = 2 * detail::offset<size_t>();
+  // sent_across_nodes
+  totalsize += detail::offset<bool>();
+  // current_time_step_id, next_time_step_id
+  totalsize += 2 * detail::offset<TimeStepId>();
+  // volume_or_ghost_mesh
+  totalsize += detail::offset<Mesh<Dim>>();
+  // interface_mesh
+  totalsize += detail::offset<Mesh<Dim - 1>>();
+  // subcell_ghost_data, dg_flux_data
+  totalsize += 2 * detail::offset<double*>();
+
+  return totalsize;
+}
+
+template <size_t Dim>
+size_t BoundaryMessage<Dim>::total_size_with_data(const size_t subcell_size,
+                                                  const size_t dg_size) {
+  size_t totalsize = total_size_without_data();
+  totalsize += (subcell_size + dg_size) * sizeof(double);
+  return totalsize;
+}
+
+template <size_t Dim>
+void* BoundaryMessage<Dim>::pack(BoundaryMessage<Dim>* in_msg) {
+  const size_t subcell_size = in_msg->subcell_ghost_data_size;
+  const size_t dg_size = in_msg->dg_flux_data_size;
+
+  const size_t totalsize = total_size_with_data(subcell_size, dg_size);
+
+  BoundaryMessage<Dim>* out_msg = reinterpret_cast<BoundaryMessage<Dim>*>(
+      CkAllocBuffer(in_msg, static_cast<int>(totalsize)));
+
+  memcpy(out_msg, &in_msg->subcell_ghost_data_size,
+         BoundaryMessage<Dim>::total_size_without_data());
+
+  if (subcell_size != 0) {
+    // double* + 1 == char* + 8 because double* is 8 bytes
+    // Place subcell data right after dg pointer
+    out_msg->subcell_ghost_data =
+        reinterpret_cast<double*>(std::addressof(out_msg->dg_flux_data)) + 1;
+    memcpy(out_msg->subcell_ghost_data, in_msg->subcell_ghost_data,
+           subcell_size * sizeof(double));
+  }
+  if (dg_size != 0) {
+    // Place dg data right after subcell data
+    out_msg->dg_flux_data =
+        reinterpret_cast<double*>(std::addressof(out_msg->dg_flux_data)) + 1 +
+        subcell_size;
+    memcpy(out_msg->dg_flux_data, in_msg->dg_flux_data,
+           dg_size * sizeof(double));
+  }
+
+  // Gotta clean up
```

Comment:
Maybe add as a comment:

A possible future optimization is if we know we have to send to another node that we define a `BoundaryMessage::allocate` function to immediate allocate a message of the right size. This will reduce the number of memory allocations when sending internode from 3 (2 on send 1 on receive) to 2 (1 on send and 1 on receive).